File: Swift/stringcalculator/NumberRemoverByIgnoredRules.swift
//
//  NumberRemoverByIgnoredRules.swift
//  stringcalculator
//
//  Created by Jorge Sánchez on 13/10/16.
//  Copyright © 2016 xurxodev. All rights reserved.
//

import Foundation

class NumberRemoverByIgnoredRules:NumberRemover {
    
    let ignoredRules:[IgnoredRule]
    
    init (ignoredRules:[IgnoredRule]){
        self.ignoredRules = ignoredRules
    }
    
    func removeNumbersToIgnore(numbers:[Int]) -> [Int]{
        
        let numbersWithoutIgnored = numbers.filter {!isIgnored($0)}
        
        return numbersWithoutIgnored;
    }
    
    private func isIgnored(number:Int) -> Bool{
        
        for ignoreRule in ignoredRules {
            if(ignoreRule.isIgnored(number)){
                return true
            }
        }

        return false;
    }
}

File: Java/src/main/java/com/xurxodev/stringcalculator/NumberRemoverByIgnoredRules.java
package com.xurxodev.stringcalculator;

import java.util.ArrayList;
import java.util.List;

public class NumberRemoverByIgnoredRules implements StringCalculator.NumberRemover{

    List<IgnoredRule> ignoredRules;

    public NumberRemoverByIgnoredRules(List<IgnoredRule> ignoredRules){
        this.ignoredRules = ignoredRules;
    }

    @Override
    public List<Integer> removeNumbersToIgnore(List<Integer> numbers) {
        List<Integer> numbersWithoutIgnored = new ArrayList<>();

        for (Integer number:numbers) {
            if (!isIgnored(number))
                numbersWithoutIgnored.add(number);
        }

        return numbersWithoutIgnored;
    }

    private boolean isIgnored(Integer number){
        for (IgnoredRule ignoreRule : ignoredRules) {
            if(ignoreRule.isIgnored(number))
                return true;
        }

        return false;
    }

    public interface IgnoredRule {

        boolean isIgnored(Integer number);

    }
}
